### DISCUSSION TITLE
Normalize newlines to `\n` in source maps

### DISCUSSION THREAD

**Discussion by Marcono1234** [webpack/webpack#16847] [Q&A] [answered]
(Disclaimer: I am not very familiar with JavaScript development and webpack, so any additional recommendations are appreciated.)

I would like to build a release artifact with webpack and include source maps inside it. However, the issue I am facing is that the source maps seem to preserve the original newlines from the source files, that is on Windows it might be `\r\n` (especially with default Git settings) whereas on Linux it will be `\n`. This prevents the release artifact from being reproducible because the source maps and therefore their checksums might differ depending on which OS you are build on.

Is there some way to configure webpack to normalize newlines in source maps to `\n`?

A workaround for this might be to use a `.gitattributes` file to force usage of `\n` regardless of OS and global Git settings, see for example the [documentation on GitHub](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings#per-repository-settings).

**Comment by alexander-akait** [comment, marked as answer]
> A workaround for this might be to use a .gitattributes file to force usage of \n regardless of OS and global Git settings, see for example the [documentation on GitHub](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings#per-repository-settings).

It is not the workaround, also you can create own plugin if you need

**Reply by Marcono1234** [reply]
> It is not the workaround

It is not? Could you please explain what you mean by that? For me that solved the issue because Git now checks out `*.ts` and `*.js` files always with `\n`, and therefore the source maps use only `\n` as well, regardless of OS. Though I still consider that a workaround because I think the cleaner solution would be if that could directly be configured for webpack.

> you can create own plugin if you need

But this is not supported at the moment by webpack, right?

**Reply by alexander-akait** [reply]
> It is not? Could you please explain what you mean by that? For me that solved the issue because Git now checks out *.ts and *.js files always with \n, and therefore the source maps use only \n as well, regardless of OS. 

You can use dependecies and they can be written with any lines, sources maps are output as is, it is `source`, i.e. as they were written

To be honest, I don't understand the problem, most likely you have already tried to find the answer, but found nothing, because it makes no sense

**Reply by alexander-akait** [reply]
And what do you mean by source maps? sourceContent inside? The source map file itself?

**Reply by Marcono1234** [reply]
> You can use dependecies and they can be written with any lines, sources maps are output as is, it is `source`, i.e. as they were written

Good point, I forget to consider that. But for my use case (creating artifact with reproducible content, regardless of OS) that does not seem to be an issue because the source maps of dependencies are downloaded during build, so their line endings are preserved (even if that happens to be `\r\n`) and therefore always consistent.

> To be honest, I don't understand the problem, most likely you have already tried to find the answer, but found nothing, because it makes no sense
> ...
> And what do you mean by source maps? sourceContent inside? The source map file itself?

To provide a specific example:
1. Download the content of https://github.com/webpack/webpack.js.org/tree/main/examples/getting-started
2. To simplify the example:
    1. Remove the `"lodash": "^4.17.21"` dependency from `package.json`
    2. Change `src/index.js` content to just:
       ```javascript
       function component() {
           const element = document.createElement("div");
           return element;
       }
       
       document.body.appendChild(component());
       ```
3. Add `devtool: 'source-map',` to `webpack.config.js`
4. Run `npm install`
5. Run `npm run build`
6. Inspect the content of `dist/main.js.map`

If you are working on Windows, cloned the repository with Git and have `core.autocrlf=true` (the default on Windows), or if you used in step 2.2 an editor which uses CRLF by default, then `src/index.js` and therefore also `dist/main.js.map` contains `\r\n` instead of just `\n`, e.g.:
```
... "sourcesContent":["function component() {\r\n  const element = document.createElement(\"div\");\r\n  return element;\r\n} ...
```

So, if you create an application and want the checksum of the `dist/main.js.map` file to be the same regardless on which machine you build the application, this might fail because on Windows the file might contain `\r\n`, whereas on Linux it will only contain `\n` and therefore have a different checksum.

In the context of this example, basically I want the source in `dist/main.js.map` to always use `\n`, regardless of whether `src/index.js` used `\n` or `\r\n`.

**Reply by alexander-akait** [reply]
Sorry it is out of scope webpack, Git is responsible for this, I said before git configuration is not a workaround, it is the right solution, you will face with the same problem and without bundler, and for any other bundlers, and even without webpack, you can try it using babel-cli and will have the same, it will be very strange if every tool do it, don't you think so?

**Reply by alexander-akait** [reply]
> In the context of this example, basically I want the source in dist/main.js.map to always use \n, regardless of whether src/index.js used \n or \r\n.

And git strongly recommends not mixing `\n` and `\r\n` between directories in the same project.

**Reply by Marcono1234** [reply]
Personally I currently think in general you should let users use whatever newline settings they want or what their code editors need, except for files which need specific line endings, e.g. LF for Linux `.sh` scripts. Being more familiar with compiled languages, I considered source code and final release artifacts two separate things, where ideally the release artifact should be reproducible. So from that perspective I would still expect that newline normalization is at least a configurable option.

Though I understand that not everyone might see it that way (maybe my view on this will change as well), and I know that supporting multiple types of newlines can complicate things, so I am fine with the `.gitattributes` solution for now since that seems to work reliably for me.

I have therefore marked this question as answered; thanks for taking the time to answer this!
